pkg/collector/corechecks/ebpf/probe/ebpfcheck/probe.go

```diff
@@ -910,3 +915,114 @@ func hashMapNumberOfEntries(mp *ebpf.Map, buffers *entryCountBuffers, maxRestart
 
 	return numElements
 }
+
+func isForEachElemHelperAvailable() bool {
+	return features.HaveProgramHelper(ebpf.SocketFilter, asm.FnForEachMapElem) == nil
+}
+
+func hashMapNumberOfEntriesWithHelper(mp *ebpf.Map) (int64, error) {
+	entryBtf := &btf.Func{
+		Name: "entry",
+		Type: &btf.FuncProto{
+			Return: &btf.Int{
+				Name:     "int",
+				Size:     4,
+				Encoding: btf.Signed,
+			},
+		},
+	}
+
+	u64Btf := &btf.Int{
+		Name:     "long",
+		Size:     8,
+		Encoding: btf.Unsigned,
+	}
+
+	callbackBtf := &btf.Func{
+		Name: "callback",
+		Type: &btf.FuncProto{
+			Return: u64Btf,
+			Params: []btf.FuncParam{
+				{
+					Name: "map",
+					Type: u64Btf,
+				},
+				{
+					Name: "key",
+					Type: u64Btf,
+				},
+				{
+					Name: "value",
+					Type: u64Btf,
+				},
+				{
+					Name: "ctx",
+					Type: u64Btf,
+				},
+			},
+		},
+	}
+
+	/*
+		equivalent to the following C code, based on the fact that `for_each_map_elem`
+		returns the amount of entries visited, so visiting all the entries ensures we
+		get the amount of entries in the map.
+
+		int callback(struct bpf_map* map, const void* key, void* value, void * ctx) {
+			return 0;
+		}
+
+		int entry() {
+			return bpf_for_each_map_elem(fd, callback, NULL, 0);
+		}
+	*/
+
+	spec := &ebpf.ProgramSpec{
```

Comment:
ah that's a very good point, we definitely could use an LRU and cache the most frequent programs

Reply:
Would we want LRU? I think it is trivial to detect when a map no longer exists and the counting program can be unloaded.

Reply:
I have been playing a bit with this and I'm less and less sure it's a good idea to cache because of the complexity of maintaining a unique key. Both fds and map id are re-used IDs, so you can never really guarantee that the program you will re-use is really for the correct map.

Regarding ignoring the programs, fully agreed. Will add something for this

Reply:
We are already relying upon our own internal map ids not changing, because we create an id->name association (since the kernel truncates names). Those associations are removed upon manager/map unload, so we could do the same.

Reply:
We also don't attempt to collect number of entries unless it is a known internal map.